module/VuFind/src/VuFind/ChannelProvider/ListItems.php

```diff
@@ -178,49 +210,61 @@ protected function buildListChannels($channelToken)
     }
 
     /**
-     * Get a list of lists, identified by ID; filter to public lists only.
+     * Get a list of public lists to display.
      *
-     * @param array $ids IDs to retrieve
+     * @param string|array $tag    Tag to match
+     * @param string|array $listId ID of list to retrieve (null for all favorites)
+     * @param string|array $userId ID of user owning favorite list
      *
      * @return array
      */
-    protected function getListsById($ids)
-    {
-        $lists = [];
-        foreach ($ids as $id) {
-            $list = $this->userList->getExisting($id);
-            if ($list->public) {
-                $lists[] = $list;
+    protected function getLists(
+        $tag = null, $listId = null, $userId = null
+    ) {
+        $resultIds = $result = [];
+
+        // Get public lists by search criteria
+        $lists = $this->resourceTags->getListsForTag($tag, $listId, $userId, true);
+
+        if ($lists->count()) {
+            foreach ($lists as $list) {
+                $resultIds[] = $list->list_id;
             }
-        }
-        return $lists;
-    }
 
-    /**
-     * Get a list of public lists to display:
-     *
-     * @return array
-     */
-    protected function getLists()
-    {
-        // First fetch hard-coded IDs:
-        $lists = $this->getListsById($this->ids);
+            $callback = function ($select) use ($resultIds) {
+                $select->where->in('id', $resultIds);
+            };
+
+            foreach ($this->userList->select($callback) as $list) {
+                $result[] = $list;
+            }
+
+            if ($listId) {
+                // Sort lists by ID list
+                $orderIds = (array)$listId;
+                $sortFn = function ($a, $b) use ($orderIds) {
+                    return
+                        array_search($a->id, $orderIds)
+                        > array_search($b->id, $orderIds);
+                };
+                usort($result, $sortFn);
+            }
+        }
 
         // Next add public lists if necessary:
         if ($this->displayPublicLists) {
-            $ids = $this->ids;
-            $callback = function ($select) use ($ids) {
+            $callback = function ($select) use ($resultIds) {
                 $select->where->equalTo('public', 1);
-                foreach ($ids as $id) {
+                foreach ($resultIds as $id) {
```

Comment:
I realize this `foreach` existed in the original code, but (unless I'm missing something) I think it would be simpler to use a `$select->where->notIn('id', $resultIds)` clause instead.